### PR TITLE
drivers: sensor: add wurth elektronik sensor drivers

### PR DESCRIPTION
**Pull request by wm-eisos**
This pull request updates the revision of the hal, removes old drivers and updates and adds new drivers.

@mah-eiSmart

### CODE REVIEW

**Comment by mah-eiSmart** [comment]
This is a big improvement to the Würth Elektronik sensor drivers, as

- the sensor names are correct now (the name without article number does not identify the sensor)
- numerous useful functions have been added
- examples are added how to use the custom attributes
- it's using the newest würth elektronik hal with the latest bug fixes

Thank you for contributing @wm-eisos and for reviewing.

BTW: The examples and drivers have been tested on several hardware platforms.

**Review by yperess — DISMISSED** [review]

**Comment by wm-eisos** [comment]
@yperess can you please rereview it.

Thanks in advance

**Comment by yperess** [comment]
> @yperess can you please rereview it.
> 
> Thanks in advance

Looks like a few extra compliance tests failing still

File: drivers/sensor/CMakeLists.txt
Comment (MaureenHelm) [inline drivers/sensor/CMakeLists.txt:61]: Please explain in the commit message why you're removing the driver

File: drivers/sensor/wsen_hids_2525020210002/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/wsen_hids_2525020210002/Kconfig:4]: Is `2525020210002` a part number or SKU? This is a long name for a driver

File: drivers/sensor/wsen_hids_2525020210002/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/wsen_hids_2525020210002/Kconfig:8]: Does this device support SPI or any other non-I2C interface? If not, you can remove the conditional if

File: drivers/sensor/wsen_hids_2525020210002/wsen_hids_2525020210002.c
Comment (MaureenHelm) [inline drivers/sensor/wsen_hids_2525020210002/wsen_hids_2525020210002.c:253]:
```suggestion
	SENSOR_DEVICE_DT_INST_DEFINE(inst, hids_2525020210002_init, NULL,                          \
				     &hids_2525020210002_data_##inst,                              \
				     &hids_2525020210002_config_##inst, POST_KERNEL,               \
				     CONFIG_SENSOR_INIT_PRIORITY, &hids_2525020210002_driver_api);
```

File: dts/bindings/sensor/we,wsen-hids-2525020210002.yaml
Comment (MaureenHelm) [inline dts/bindings/sensor/we,wsen-hids-2525020210002.yaml:15]: The default value needs to be justified in the description. For example, is it a hardware register reset value?

File: dts/bindings/sensor/we,wsen-hids-2525020210002.yaml
Comment (MaureenHelm) [inline dts/bindings/sensor/we,wsen-hids-2525020210002.yaml:26]: Same here

File: samples/sensor/wsen_hids_2525020210002/README.rst
Comment (MaureenHelm) [inline samples/sensor/wsen_hids_2525020210002/README.rst:15]: We're trying to stop introducing new driver-specific samples and move toward having more generic samples for each sensor type. Please look at `samples/sensor/accel_polling` for example

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
This would be easier to review and get merged if you break it up into multiple PRs. Each PR should just cover one driver, removing the old one and adding the new one.

File: drivers/sensor/wsen_hids_2525020210002/Kconfig
Comment (wm-eisos) [reply drivers/sensor/wsen_hids_2525020210002/Kconfig:4]: This is a part number uniquely identifying the sensor as for example hids refers to the family of the sensor (humidity in this case), so it is important to have it part of the driver name.

File: dts/bindings/sensor/we,wsen-hids-2525020210002.yaml
Comment (wm-eisos) [reply dts/bindings/sensor/we,wsen-hids-2525020210002.yaml:15]: default value explanation has been added.

File: dts/bindings/sensor/we,wsen-hids-2525020210002.yaml
Comment (wm-eisos) [reply dts/bindings/sensor/we,wsen-hids-2525020210002.yaml:28]: default value explanation has been added.

File: drivers/sensor/wsen_hids_2525020210002/wsen_hids_2525020210002.c
Comment (wm-eisos) [reply drivers/sensor/wsen_hids_2525020210002/wsen_hids_2525020210002.c]: this has been applied to all of the drivers.

File: samples/sensor/wsen_hids_2525020210002/README.rst
Comment (wm-eisos) [reply samples/sensor/wsen_hids_2525020210002/README.rst:15]: The samples for the drivers has been supplied because some of the drivers implement custom attribues or triggers. which without looking at the example would be really hard for the user to use.

**Comment by wm-eisos** [comment]
> This would be easier to review and get merged if you break it up into multiple PRs. Each PR should just cover one driver, removing the old one and adding the new one.

They have all been done in one PR since in this PR the halhas been updated with breaking changes to the old drivers and the new drivers need the new hal to function corectly.

**Comment by wm-eisos** [comment]
@yperess i see that a couple of checks have been stuck for a while now. Do these require a review before they trigger?

File: drivers/sensor/CMakeLists.txt
Comment (MaureenHelm) [reply drivers/sensor/CMakeLists.txt:61]: Would like to see more explanation than just "we're removing the driver to replace it with a new one". You mention the reason in https://github.com/zephyrproject-rtos/zephyr/pull/62408#issuecomment-1711566468, but would like to have that captured in the commit history.

File: samples/sensor/wsen_hids_2525020210002/README.rst
Comment (MaureenHelm) [reply samples/sensor/wsen_hids_2525020210002/README.rst:15]: Can we postpone the new samples to another PR?

**Comment by wm-eisos** [comment]
> > @yperess i see that a couple of checks have been stuck for a while now. Do these require a review before they trigger?
> 
> Approved the workflow

Can you please reapprove it again now, i rann the check compliance py script this time and solved the issues hopefully no more compliance check fails.

File: samples/sensor/wsen_hids_2525020210002/README.rst
Comment (wm-eisos) [reply samples/sensor/wsen_hids_2525020210002/README.rst]: i just removed them from this PR and will create a seperate one affter this one is merged.

File: drivers/sensor/CMakeLists.txt
Comment (wm-eisos) [reply drivers/sensor/CMakeLists.txt:62]: i changed the commit message to indicate that the new hal is not compatible with the old drivers.

**Comment by wm-eisos** [comment]
@yperess @MaureenHelm since all checks have passed this time. can you please review it before any merge conflicts could happen.
Thanks in advance

**Comment by wm-eisos** [comment]
@yperess @MaureenHelm this pull request is now ready for review

**Comment by wm-eisos** [comment]
@MaureenHelm @yperess i just resolved the merge conflicts again. please review the commits before another merge conflict occurs.

**Comment by wm-eisos** [comment]
@yperess can you please approve the workflow

**Comment by wm-eisos** [comment]
Should i open another PR with the same commits as it seems to appear that this pull request has been inactive for 2 weeks

@MaureenHelm @yperess @nashif @teburd  @tristan-google

**Comment by kartben** [comment]
> Should i open another PR with the same commits as it seems to appear that this pull request has been inactive for 2 weeks
> 
> @MaureenHelm @yperess @nashif @teburd  @tristan-google 

I've triggered the workflow, there should be no need to re-open a PR since you already did so yourself. FYI given we are now in feature freeze for 3.5 this won't be another week until it can be considered for merging though.

**Comment by wm-eisos** [comment]
@kartben can you please retrigger it

File: drivers/sensor/wsen_hids_2525020210002/wsen_hids_2525020210002.c
Comment (MaureenHelm) [inline drivers/sensor/wsen_hids_2525020210002/wsen_hids_2525020210002.c:202]: Need to check if the I2C bus is ready with  `i2c_is_ready_dt` before using it.

File: drivers/sensor/CMakeLists.txt
Comment (MaureenHelm) [inline drivers/sensor/CMakeLists.txt]: Please keep these in alpha order

File: drivers/sensor/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/Kconfig]: Please keep these in alpha order

File: drivers/sensor/wsen_pdus_25131308XXXXX/wsen_pdus_25131308XXXXX.c
Comment (MaureenHelm) [inline drivers/sensor/wsen_pdus_25131308XXXXX/wsen_pdus_25131308XXXXX.c:64]: Are the magic numbers here and throughout this function documented somewhere?

File: drivers/sensor/wsen_pdus_25131308XXXXX/wsen_pdus_25131308XXXXX.c
Comment (MaureenHelm) [inline drivers/sensor/wsen_pdus_25131308XXXXX/wsen_pdus_25131308XXXXX.c:129]: Need to call `i2c_is_ready_dt`

File: dts/bindings/sensor/we,wsen-pdus-25131308XXXXX.yaml
Comment (MaureenHelm) [inline dts/bindings/sensor/we,wsen-pdus-25131308XXXXX.yaml]: Move the details into the description field so it will get rendered in the [documentation](https://docs.zephyrproject.org/latest/build/dts/api/bindings.html)

File: drivers/sensor/wsen_tids_2521020222501/wsen_tids_2521020222501.c
Comment (MaureenHelm) [inline drivers/sensor/wsen_tids_2521020222501/wsen_tids_2521020222501.c:152]: Please no camel case

File: drivers/sensor/wsen_tids_2521020222501/wsen_tids_2521020222501_trigger.c
Comment (MaureenHelm) [inline drivers/sensor/wsen_tids_2521020222501/wsen_tids_2521020222501_trigger.c:56]: Need to save a pointer to the user's sensor trigger and pass it back, rather than create a new sensor trigger on the stack. See #55130

File: drivers/sensor/wsen_pads_2511020213301/wsen_pads_2511020213301_trigger.c
Comment (MaureenHelm) [inline drivers/sensor/wsen_pads_2511020213301/wsen_pads_2511020213301_trigger.c:54]: Don't create new trigger structs on the stack; store a pointer to the user's sensor trigger and pass it back

File: drivers/sensor/wsen_itds_2533020201601/wsen_itds_2533020201601_trigger.c
Comment (MaureenHelm) [inline drivers/sensor/wsen_itds_2533020201601/wsen_itds_2533020201601_trigger.c:73]: Don't create new trigger structs on the stack; store a pointer to the user's sensor trigger and pass it back

File: tests/drivers/build_all/sensor/sensors_trigger_global.conf
Comment (MaureenHelm) [inline tests/drivers/build_all/sensor/sensors_trigger_global.conf:60]: Please keep alpha order

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/wsen_hids_2525020210002/wsen_hids_2525020210002.c
Comment (wm-eisos) [reply drivers/sensor/wsen_hids_2525020210002/wsen_hids_2525020210002.c:206]: For every driver now the correct is ready function is called depending on the interface

File: drivers/sensor/CMakeLists.txt
Comment (wm-eisos) [reply drivers/sensor/CMakeLists.txt]: Fixed

File: drivers/sensor/wsen_pdus_25131308XXXXX/wsen_pdus_25131308XXXXX.c
Comment (wm-eisos) [reply drivers/sensor/wsen_pdus_25131308XXXXX/wsen_pdus_25131308XXXXX.c]: I have added a comment that says that these value are mentioned in the user manual

File: drivers/sensor/wsen_pdus_25131308XXXXX/wsen_pdus_25131308XXXXX.c
Comment (wm-eisos) [reply drivers/sensor/wsen_pdus_25131308XXXXX/wsen_pdus_25131308XXXXX.c]: Fixed

File: drivers/sensor/wsen_tids_2521020222501/wsen_tids_2521020222501.c
Comment (wm-eisos) [reply drivers/sensor/wsen_tids_2521020222501/wsen_tids_2521020222501.c]: I made all variables of the form lower case with underscores between different nouns

File: drivers/sensor/wsen_tids_2521020222501/wsen_tids_2521020222501_trigger.c
Comment (wm-eisos) [reply drivers/sensor/wsen_tids_2521020222501/wsen_tids_2521020222501_trigger.c]: The trigger pointer is now stored

**Comment by wm-eisos** [comment]
please retrigger the workflow i just marked the PR as ready for review

**Comment by wm-eisos** [comment]
@MaureenHelm any updates regarding the rereview?

**Comment by wm-eisos** [comment]
Latest force push was to apply 191865b on the drivers that have triggers

**Comment by wm-eisos** [comment]
@kartben @MaureenHelm can you please review the changes?

**Comment by mah-eiSmart** [comment]
Hello @carlescufi , hello @MaureenHelm ,

please approve this pull request. People are waiting to get access to the new version of the drivers.

Thank you for the support.

**Comment by MaureenHelm** [comment]
@teburd @yperess can I get your help reviewing this please?

**Comment by wm-eisos** [comment]
@MaureenHelm any updates regarding this?

**Comment by wm-eisos** [comment]
@kartben @yperess can you guys please check out this PR?

**Review by yperess — DISMISSED** [review]

**Comment by wm-eisos** [comment]
@MaureenHelm can you please check it out. im not sure if i should rebase,resolve the conflict and force push again.

**Comment by wm-eisos** [comment]
@kartben any updates please?

File: samples/sensor/wsen_hids/README.rst
Comment (kartben) [inline samples/sensor/wsen_hids/README.rst:1]: All the removed samples need entries added to `doc\_scripts\redirects.py` so that e.g. https://docs.zephyrproject.org/3.5.0/samples/sensor/wsen_hids/README.html takes people to a page that exists when they click on "The latest development version of this page may be more current than this released x.y.z version."
You'll want to redirect to either samples/sensor or whichever "generic" sample makes the most sense for each.

**Comment by kartben** [comment]
> @MaureenHelm @kartben can you please check it out. im not sure if i should rebase, resolve the conflict and force push again.

This will dismiss existing reviews but yes, this will need to be done eventually, i.e new sensors starting at 0x81 at the time of writing

**Comment by wm-eisos** [comment]
> > @MaureenHelm @kartben can you please check it out. im not sure if i should rebase, resolve the conflict and force push again.
> 
> This will dismiss existing reviews but yes, this will need to be done eventually, i.e new sensors starting at 0x81 at the time of writing

any solution regarding the times to get the PRs approved it has been a cycle of me rebasing and fixing conflicts and before the PR is merged there is already another conflict

**Comment by kartben** [comment]
> > > @MaureenHelm @kartben can you please check it out. im not sure if i should rebase, resolve the conflict and force push again.
> > 
> > 
> > This will dismiss existing reviews but yes, this will need to be done eventually, i.e new sensors starting at 0x81 at the time of writing
> 
> any solution regarding the times to get the PRs approved it has been a cycle of me rebasing and fixing conflicts and before the PR is merged there is already another conflict

Not really a criticism of your approach here, maybe there was no other way than to lump everything under the same PR (?), but frankly I think this boils down to the change being thousands of lines, which is incredibly much harder to manage for everyone involved, including reviewers.

See https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow -- "One general practice we encourage, is to make small, controlled changes. This practice simplifies review, makes merging and rebasing easier, and keeps the change history clear and clean."

For example, I am about to provide you with a few comments of things that need fixing, but as you'll have addressed them, that still won't be enough for me to give a +1 as the rest of the PR is beyond my domain of expertise.

**Comment by wm-eisos** [comment]
> > > > @MaureenHelm @kartben can you please check it out. im not sure if i should rebase, resolve the conflict and force push again.
> > > 
> > > 
> > > This will dismiss existing reviews but yes, this will need to be done eventually, i.e new sensors starting at 0x81 at the time of writing
> > 
> > 
> > any solution regarding the times to get the PRs approved it has been a cycle of me rebasing and fixing conflicts and before the PR is merged there is already another conflict
> 
> Not really a criticism of your approach here, maybe there was no other way than to lump everything under the same PR (?), but frankly I think this boils down to the change being thousands of lines, which is incredibly much harder to manage for everyone involved, including reviewers.
> 
> See https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow -- "One general practice we encourage, is to make small, controlled changes. This practice simplifies review, makes merging and rebasing easier, and keeps the change history clear and clean."
> 
> For example, I am about to provide you with a few comments of things that need fixing, but as you'll have addressed them, that still won't be enough for me to give a +1 as the rest of the PR is beyond my domain of expertise.

The thing is in this PR there is a hal update which produces breaking changes for the older drivers. Thats why i have mutliple drivers in the same PR.

File: dts/bindings/sensor/we,wsen-itds.yaml
Comment (kartben) [inline dts/bindings/sensor/we,wsen-itds.yaml:1]: If you're dropping / renaming drivers, this needs to be documented in the release notes and migration guide as this will be requiring users update their code, and they need to know what to do

**Review by kartben — CHANGES_REQUESTED** [review]

File: dts/bindings/sensor/we,wsen-itds.yaml
Comment (wm-eisos) [reply dts/bindings/sensor/we,wsen-itds.yaml:1]: Is there a guide on which files need to be modified?

**Comment by kartben** [comment]
> The thing is in this PR there is a hal update which produces breaking changes for the older drivers. 

So are you saying that just taking https://github.com/zephyrproject-rtos/zephyr/pull/62408/commits/fd5cecd7b4168079c3c2ecdc5784fc728832d43f leads to a configuration where drivers don't compile? We try to keep things bissectable so it would be good to fix if needed.

> Thats why i have mutliple drivers in the same PR.

Yup but for example dropping the "non-generic" samples could be an entirely different PR, completely independent of the HAL bump, right? And this one would likely take a lot less longer to review.

File: dts/bindings/sensor/we,wsen-itds.yaml
Comment (kartben) [reply dts/bindings/sensor/we,wsen-itds.yaml:1]: doc/releases/release-notes-3.6.rst
doc/releases/migration-guide-3.6.rst

**Comment by kartben** [comment]
> Regarding the first point that is correct if the commit is merged then yes the older drivers wont compile.

Then please squash accordingly so that all commits correspond to a working state.

**Comment by wm-eisos** [comment]
> > Regarding the first point that is correct if the commit is merged then yes the older drivers wont compile.
> 
> Then please squash accordingly so that all commits correspond to a working state.

what exactly should i squash

**Comment by wm-eisos** [comment]
depends on PR (#68462) to be merged

**Comment by wm-eisos** [comment]
@kartben regarding this PR, it is currently in draft but since the commits are no longer in this PR. This PR will fail because it still has the examples for the removed drivers? Anything to do regarding that ? i would like both to be merged for RC1.

**Comment by kartben** [comment]
@wm-eisos fyi and just to set expectations, we have entered feature freeze for Zephyr 3.6 release so it might be simpler to hold off for now ; in particular, the migration guide for 3.7 doesn't exist just yet so it will be easier for you to add to it once it's been seeeded after 3.6 is out. Thanks!

**Comment by wm-eisos** [comment]
any updates regarding this PR ?
@kartben @yperess @aescolar

**Comment by mah-eiSmart** [comment]
> This pull request has been marked as stale because it has been open (more than) 60 days with no activity. Remove the stale label or add a comment saying that you would like to have the label removed otherwise this pull request will automatically be closed in 14 days. Note, that you can always re-open a closed pull request at any time.

Please remove the stale label. We are still waiting for review and merge.

**Comment by aescolar** [comment]
> Please remove the stale label. We are still waiting for review and merge.

@mah-eiSmart You will need to rebase the PR as it is currently conflicting with main.
@MaureenHelm, @kartben could you take another look?

**Comment by aescolar** [comment]
@wm-eisos ^^

**Comment by mah-eiSmart** [comment]
> > Please remove the stale label. We are still waiting for review and merge.
> 
> @mah-eiSmart You will need to rebase the PR as it is currently conflicting with main. @MaureenHelm, @kartben could you take another look?

@aescolar , yes that's right. My colleague, which is back next week, will do this. Actually, as you see in the history, we rebased already many times. After rebase, it would be nice to get finally the approval before the next incompatible changes of the master make the review again void.

**Comment by wm-eisos** [comment]
> @wm-eisos ^^

I have rebased the PR yesterday. please try to take another look as soon as possible.

**Comment by aescolar** [comment]
@MaureenHelm, @kartben could you take another look at this PR?

**Review by decsny — DISMISSED** [review]
are all of these bindings really needed at this point

**Comment by wm-eisos** [comment]
> are all of these bindings really needed at this point

Yes as these bindings are required to setup the sensor acccording to the behaviour that the user wants. PS this PR contains drivers for multiple sensors 7 in total

**Comment by wm-eisos** [comment]
Can you please remove the Stale label, i have rebased the PR can you please check it as well.

**Comment by decsny** [comment]
there is a merge conflict

**Comment by wm-eisos** [comment]
> there is a merge conflict

I will solve it now however as you can see in this PR this loop keeps hapening where the merge conflicts are resolved and then new merge conflicts come up because the PR is still not merged

**Comment by decsny** [comment]
> > there is a merge conflict
> 
> I will solve it now however as you can see in this PR this loop keeps hapening where the merge conflicts are resolved and then new merge conflicts come up because the PR is still not merged

welcome to my life as well

**Comment by decsny** [comment]
@mah-eiSmart it looks like you approved of this PR, is that still the case

File: dts/bindings/sensor/we,wsen-hids-2525020210001-i2c.yaml
Comment (decsny) [inline dts/bindings/sensor/we,wsen-hids-2525020210001-i2c.yaml:1]: can you maybe put all these bindings in a subfolder, it's pretty noisy to have all of these in the top sensor binding folder

**Comment by mah-eiSmart** [comment]
> @mah-eiSmart it looks like you approved of this PR, is that still the case

Yes, from my side it's still fine!

File: dts/bindings/sensor/we,wsen-hids-2525020210001-i2c.yaml
Comment (wm-eisos) [reply dts/bindings/sensor/we,wsen-hids-2525020210001-i2c.yaml:1]: As i currently see no other sensor goups its bindings also should i group the bindings per sensor or per vendor. And if you have an example ffor a sensor with grouped bindings that would be great

Comment (decsny) [reply dts/bindings/sensor/we,wsen-hids-2525020210001-i2c.yaml:1]: it doesn't matter if there's no examples like that, and the structure of the bindings folders doesnt functionally affect anything, you could move these bindings to the wifi binding folder and it would not make any difference except to humans looking at the folders being confused why that would be there. so, it would help to limit the noise by putting these associated bindings in a subfolder IMO

Comment (decsny) [reply dts/bindings/sensor/we,wsen-hids-2525020210001-i2c.yaml:1]: and to answer your question, i think you should group it by sensor, not vendor

Comment (wm-eisos) [reply dts/bindings/sensor/we,wsen-hids-2525020210001-i2c.yaml:1]: okay will do

**Review by decsny — APPROVED** [review]
somebody with sensor knowledge should review

**Comment by dkalowsk** [comment]
> somebody with sensor knowledge should review

@MaureenHelm as sensor maintainer, are you able to take a review of this?

**Comment by mah-eiSmart** [comment]
> > somebody with sensor knowledge should review
> 
> @MaureenHelm as sensor maintainer, are you able to take a review of this?

@dkalowsk Seems, @MaureenHelm is busy. Can we go forward without her help?

**Comment by dkalowsk** [comment]
> > > somebody with sensor knowledge should review
> > 
> > 
> > @MaureenHelm as sensor maintainer, are you able to take a review of this?
> 
> @dkalowsk Seems, @MaureenHelm is busy. Can we go forward without her help?

There are ways to do so.  You can escalate in the developer meeting or on discord. I’m a little more concerned that this PR is a year in the making and stalling out on reviews. @nashif @MaureenHelm ping.

**Comment by mah-eiSmart** [comment]
> > > > somebody with sensor knowledge should review
> > > 
> > > 
> > > @MaureenHelm as sensor maintainer, are you able to take a review of this?
> > 
> > 
> > @dkalowsk Seems, @MaureenHelm is busy. Can we go forward without her help?
> 
> There are ways to do so. You can escalate in the developer meeting or on discord. I’m a little more concerned that this PR is a year in the making and stalling out on reviews. @nashif @MaureenHelm ping.

@dkalowsk , this PR has been adopted to new versions of Zephyr many times, in the past 12 months. Thus it's still up to date. Yesterday I got feedback that customers did not choose that sensor products, because the drivers are not up to date. Escalating on discord or on developer meeting we did not do so far, we have no experience with it.

**Comment by wm-eisos** [comment]
This PR will be closed as the plan is to have seperate PRs per sensor

**Pull request closed without merging**
